### PR TITLE
change blockstore data structure to support lock-free read access

### PR DESCRIPTION
**Pull request by asiegel-jt**

### CODE REVIEW

File: src/flamenco/runtime/fd_blockstore.h [hunk 1]
```diff
@@ -165,32 +166,29 @@ struct fd_block {
   Note random access of individual shred indices is not performant, due to the variable-length
   nature of shreds. */
 
+  ulong tot_sz;       /* total size of block allocation */
   ulong data_gaddr;   /* ptr to the beginning of the block's allocated data region */
   ulong data_sz;      /* block size */
   ulong shreds_gaddr; /* ptr to the list of fd_blockstore_shred_t */
   ulong shreds_cnt;
-  ulong micros_gaddr; /* ptr to the list of fd_blockstore_micro_t */
-  ulong micros_cnt;
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:173]: don't remove this, as we'll need it for executing microblocks in parallel

File: src/flamenco/runtime/fd_blockstore.h [hunk 2]
```diff
@@ -165,32 +166,29 @@ struct fd_block {
   Note random access of individual shred indices is not performant, due to the variable-length
   nature of shreds. */
 
+  ulong tot_sz;       /* total size of block allocation */
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:169]: block_sz

File: src/flamenco/runtime/fd_blockstore.h [hunk 3]
```diff
@@ -476,10 +468,26 @@ fd_blockstore_block_frontier_query( fd_blockstore_t * blockstore,
                                     ulong *           parents,
                                     ulong             parents_sz );
 
+/* Query the block data and meta in a thread-safe manner which will
+   not block writes. The data and meta is copied out into memory
+   allocated with the given alloc. A NULL is returned on error. */
+uchar *
+fd_blockstore_block_query_safe( fd_blockstore_t * blockstore, ulong slot, fd_valloc_t alloc, fd_block_t * blk_out, fd_slot_meta_t * slot_meta_out, ulong * data_sz_out );
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:475]: caller should manage and pass in the memory, can't assume they have an alloc

Comment (lidatong) [reply src/flamenco/runtime/fd_blockstore.h:475]: i think `query_volatile` is a better name, `safe` doesn't really communicate threads and also could mislead the user as you can get torn write data even if it doesn't crash

File: src/flamenco/runtime/fd_blockstore.h [hunk 4]
```diff
@@ -476,10 +468,26 @@ fd_blockstore_block_frontier_query( fd_blockstore_t * blockstore,
...
+
+/* Query the block meta data in a thread-safe manner which will
+   not block writes. The meta is copied out. */
+int
+fd_blockstore_meta_query_safe( fd_blockstore_t * blockstore, ulong slot, fd_block_t * blk_out, fd_slot_meta_t * slot_meta_out );
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:480]: can you make meta more explicit that it's `slot_meta_query_volatile`?

File: src/flamenco/runtime/fd_blockstore.h [hunk 5]
```diff
@@ -476,10 +468,26 @@ fd_blockstore_block_frontier_query( fd_blockstore_t * blockstore,
...
+
 /* Query the transaction data for the given signature */
 fd_blockstore_txn_map_t *
 fd_blockstore_txn_query( fd_blockstore_t * blockstore, uchar const sig[FD_ED25519_SIG_SZ] );
 
+/* Query the transaction data for the given signature in a thread
+ * safe manner */
+int
+fd_blockstore_txn_query_safe( fd_blockstore_t * blockstore, uchar const sig[FD_ED25519_SIG_SZ], fd_blockstore_txn_map_t * txn_out, long * blk_ts, uchar txn_data_out[FD_TXN_MTU] );
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.h:489]: add `static` in the array decl

File: src/flamenco/runtime/fd_blockstore.c [hunk 6]
```diff
@@ -853,79 +828,191 @@ fd_blockstore_shred_query_copy_data( fd_blockstore_t * blockstore, ulong slot, u
...
+    fd_block_t * blk = fd_wksp_laddr_fast( wksp, blk_gaddr );
+    fd_memcpy( blk_out, blk, sizeof(fd_block_t) );
+    ulong ptr = blk->data_gaddr;
+    if( FD_UNLIKELY( !ptr || ptr == ULONG_MAX ) ) return NULL;
+    ulong sz = *data_sz_out = blk->data_sz;
+    if( sz >= FD_SHRED_MAX_PER_SLOT * FD_SHRED_MAX_SZ ) continue;
+
+    if( FD_UNLIKELY( fd_readwrite_check_concur_read( &blockstore->lock, seqnum ) ) ) continue;
+
+    uchar * data_out = fd_valloc_malloc( alloc, 8UL, sz );
```
Comment (lidatong) [inline src/flamenco/runtime/fd_blockstore.c:914]: is this right? data is 8-byte aligned?

File: src/flamenco/runtime/fd_blockstore.h (hunk 3 above)
Comment (asiegel-jt) [reply src/flamenco/runtime/fd_blockstore.h:475]: the caller doesn't know the size of the block, and there is no safe way to read it ahead of time

File: src/flamenco/runtime/fd_blockstore.h (hunk 5 above)
Comment (asiegel-jt) [reply src/flamenco/runtime/fd_blockstore.h]: txn_data_out can be NULL

**Review by lidatong — APPROVED** [review]

**Pull request merged**